// libs/remoteApi.js
var http = require( 'http' );
var qs = require( 'querystring' );
var api = require( '../config/host.config.json' );
var fs = require('fs');

// post请求
	var post = function( req , remoteUrl ,  success ) {
	  var body = qs.stringify( req.body );
	  var method = req.method;
	  var options = {
		  hostname: api.host,
		  port: api.port,  
		  path: remoteUrl,  
		  method: 'post',
		  headers: req.headers
	  }

	  console.log("host:%s -- url:%s -- method:%s -- port:%d" , api.host , remoteUrl , 'post' , api.port);

	  var req = http.request(options, function(res) {
				res.setEncoding('utf8');
				var data = '';
				res.on('data', function (d, err) {
					if( err ) {
					  console.error( '请求数据失败' );
					  return;
					}
				  data += d;
				}).on('end' , function(){
						
					data = JSON.parse( data );
					success( res , data );
				});
			});

		req.on('error', function(e){
		   console.log("auth_user error: " + e.message);
		});
		req.write( body );

		req.end();
	}


exports.post = post;
